libs/wingc/src/lsp/hover.rs

```diff
@@ -0,0 +1,122 @@
+use lsp_types::{Hover, HoverContents, MarkupContent, MarkupKind};
+
+use crate::lsp::ast_traversal::find_symbol;
+use crate::lsp::sync::FILES;
+use crate::{
+	ast::ExprKind,
+	wasm_util::{ptr_to_string, string_to_combined_ptr},
+};
+
+#[no_mangle]
+pub unsafe extern "C" fn wingc_on_hover(ptr: u32, len: u32) -> u64 {
+	let parse_string = ptr_to_string(ptr, len);
+	if let Ok(parsed) = serde_json::from_str(&parse_string) {
+		if let Some(token_result) = on_hover(parsed) {
+			let result = serde_json::to_string(&token_result).unwrap();
+
+			string_to_combined_ptr(result)
+		} else {
+			0
+		}
+	} else {
+		eprintln!("Failed to parse 'onHover' text document: {}", parse_string);
+		0
+	}
+}
+pub fn on_hover<'a>(params: lsp_types::HoverParams) -> Option<Hover> {
+	FILES.with(|files| {
+		let files = files.borrow();
+		let files = files.read().unwrap();
+		let parse_result = files.get(&params.text_document_position_params.text_document.uri.clone());
+		let parse_result = parse_result.unwrap();
+
+		let position = params.text_document_position_params.position;
+
+		let root_scope = &parse_result.scope;
+		let root_env_ref = root_scope.env.borrow();
+		let root_env = root_env_ref.as_ref().unwrap();
+
+		if let Some(symbol_context) = find_symbol(root_scope, &position) {
+			let symbol = symbol_context.1;
+			let expr = symbol_context.0.expression;
+			let scope = symbol_context.0.scope;
+			let symbol_name = &symbol.name;
+			let expression_type = if let Some(expr) = expr {
+				let t = expr.evaluated_type.borrow();
+				t.clone()
+			} else {
+				None
+			};
+			let reference = if let Some(expr) = expr {
+				if let ExprKind::Reference(reference) = &expr.kind {
+					Some(reference)
+				} else {
+					None
+				}
+			} else {
+				None
+			};
```

Comment:
I think this is slightly more idiomatic rust
```suggestion
			let expression_type = expr.as_ref().and_then(|expr| {
				let t = expr.evaluated_type.borrow();
				t.clone()
			});
			let reference = expr.as_ref().and_then(|expr| {
				if let ExprKind::Reference(reference) = &expr.kind {
					Some(reference)
				} else {
					None
				}
			});
```